functorch/experimental/_map.py

```diff
@@ -17,23 +18,77 @@
     _get_current_dispatch_mode,
     _pop_mode_temporarily,
 )
-from torch.utils._pytree import tree_flatten
 from ._cond import _has_potential_branch_input_alias, _has_potential_branch_input_mutation, UnsupportedAliasMutationException
 
 
-map = HigherOrderOperator("map")
-
-
-def trace_map(proxy_mode, func_overload, f, xs, *args):
-    if not isinstance(xs, torch.Tensor):
-        raise ValueError("map() must loop over a tensor")
-    if len(xs.shape) == 0 or xs.shape[0] == 0:
-        raise ValueError("map() cannot be traced with scalar tensors or zero dimension tensors")
-    if not all(isinstance(o, torch.Tensor) for o in args):
-        raise ValueError("map() operands must be a list of tensors or modules")
+# TODO: We add this to prevent dymamo tracing into map_wrapper,
+# remove the wrapper call when it's ready.
+class MapWrapper(HigherOrderOperator):
+    def __call__(self, *args, **kwargs):
+        return map_wrapper(*args, **kwargs)
+
+map = MapWrapper("map")
+map_impl = HigherOrderOperator("map_impl")
+
+def map_wrapper(f, xs, *args):
+    flat_xs, xs_spec = pytree.tree_flatten(xs)
+    assert all(isinstance(t, torch.Tensor) for t in flat_xs), f"mapped xs can only contain tensors got {flat_xs}"
+
+    num_mapped_args = len(flat_xs)
+    assert num_mapped_args > 0, "map must have at least one mapped argument."
+
+    shapes = [xs.shape for xs in flat_xs]
+    leading_dim_size = shapes[0][0]
+    assert len(shapes) > 0 and all(cur_shape[0] == leading_dim_size for cur_shape in shapes)
+
+    out_spec = [None]
+
+    def flat_fn(num_mapped_args, *flat_args):
+        xs = pytree.tree_unflatten(flat_args[:num_mapped_args], xs_spec)
+        unflattened_out = f(xs, *flat_args[num_mapped_args:])
+        flat_out, tmp_out_spec = pytree.tree_flatten(unflattened_out)
+        out_spec[0] = tmp_out_spec
+        return flat_out
+    return pytree.tree_unflatten(map_impl(flat_fn, num_mapped_args, *flat_xs, *args), out_spec[0])
+
+class MapAutogradOp(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, fw_graph, joint_graph, num_mapped_args, *flat_args):
+        ctx.save_for_backward(*flat_args)
+        ctx._joint_graph = joint_graph
+        ctx._num_mapped_args = num_mapped_args
+        _ = torch._C._AutoDispatchBelowAutograd()
```

Comment:
This should be in a try-finally block. If an exception happens, Python may hold onto the object instead of releasing it. (RAII isn't really a thing in Python, but it is fine as long as we don't create reference cycles and do try-finally)

Reply:
Ok!

Reply:
@zou3519
Out of curiosity, how are well implemented context managers semantically different from RAII? (Surprised `_AutoDispatchBelowAutograd` isn't implemented as a context manager if that is the case).